src/main/java/com/aws/greengrass/componentmanager/plugins/DefaultDockerClient.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.componentmanager.plugins;
+
+import com.aws.greengrass.componentmanager.plugins.exceptions.ImagePullRetryableException;
+import com.aws.greengrass.componentmanager.plugins.exceptions.ImagePullServiceException;
+import com.aws.greengrass.componentmanager.plugins.exceptions.InvalidImageOrAccessDeniedException;
+import com.aws.greengrass.componentmanager.plugins.exceptions.UserNotAuthorizedForDockerException;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.util.Exec;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * Docker CLI wrapper that communicates with Docker Engine to execute user commands.
+ */
+@NoArgsConstructor
+public class DefaultDockerClient {
+    public static final Logger logger = LogManager.getLogger(DefaultDockerClient.class);
+
+    /**
+     * Sanity check for installation.
+     *
+     * @return if docker is installed on the host
+     */
+    public boolean dockerInstalled() {
+        CliResponse response = runDockerCmd("docker -v");
+        return response.exit.isPresent() && response.exit.get() == 0;
+    }
+
+    /**
+     * Login to given docker registry.
+     *
+     * @param registry Registry to log into, with credentials encapsulated
+     * @return if docker login was successful
+     * @throws ImagePullServiceException           error in authenticating with the registry
+     * @throws UserNotAuthorizedForDockerException when current user is not authorized to use docker
+     * @throws IOException                         unexpected error
+     */
+    public boolean login(Registry registry)
+            throws ImagePullServiceException, UserNotAuthorizedForDockerException, IOException {
+        // TODO : [Blocker] Ensure credentials in command to be run do not get logged
+        CliResponse response = runDockerCmd(String.format("docker login %s -u %s -p %s", registry.getEndpoint(),
```

Comment:
If you're looking for programmatic replacement for CLI, unfortunately there's [no official Java SDK](https://docs.docker.com/engine/api/sdk/#unofficial-libraries). We're using https://github.com/docker-java/docker-java for in UAT framework which is a good candidate.

Reply:
Discussed offline about this